src/catalog/catalog.cpp

```diff
@@ -734,40 +617,129 @@ storage::Database *Catalog::GetDatabaseWithName(
+    throw CatalogException("Database " + std::to_string(database_oid) +
+                           " is not found");
+  }
+
+  return database_object;
+}
+
+/* Check table from pg_table with table_name using txn,
+ * get it from storage layer using table_oid,
+ * throw exception and abort txn if not exists/invisible
+ * */
+std::shared_ptr<TableCatalogObject> Catalog::GetTableObject(
+    const std::string &database_name, const std::string &table_name,
+    concurrency::Transaction *txn) {
+  if (txn == nullptr) {
+    throw CatalogException("Do not have transaction to get table object " +
+                           database_name + "." + table_name);
   }
 
   LOG_TRACE("Looking for table %s in database %s", table_name.c_str(),
             database_name.c_str());
 
   // Check in pg_database, throw exception and abort txn if not exists
-  auto database_oid =
-      DatabaseCatalog::GetInstance()->GetDatabaseOid(database_name, txn);
+  auto database_object =
+      DatabaseCatalog::GetInstance()->GetDatabaseObject(database_name, txn);
 
-  if (database_oid == INVALID_OID) {
+  if (!database_object || database_object->database_oid == INVALID_OID) {
     throw CatalogException("Database " + database_name + " is not found");
   }
 
   // Check in pg_table using txn
-  auto table_oid =
-      TableCatalog::GetInstance()->GetTableOid(table_name, database_oid, txn);
+  auto table_object = database_object->GetTableObject(table_name);
 
-  if (table_oid == INVALID_OID) {
+  if (!table_object || table_object->table_oid == INVALID_OID) {
     // throw table not found exception and explicitly abort txn
     throw CatalogException("Table " + table_name + " is not found");
   }
 
-  if (single_statement_txn) {
-    txn_manager.CommitTransaction(txn);
+  return table_object;
+}
+
+std::shared_ptr<TableCatalogObject> Catalog::GetTableObject(
+    oid_t database_oid, oid_t table_oid, concurrency::Transaction *txn) {
+  if (txn == nullptr) {
+    throw CatalogException("Do not have transaction to get table object " +
+                           std::to_string(database_oid) + "." +
+                           std::to_string(table_oid));
   }
-  return storage_manager->GetTableWithOid(database_oid, table_oid);
+
+  LOG_TRACE("Looking for table %u in database %u", table_oid, database_oid);
+
+  // Check in pg_database, throw exception and abort txn if not exists
+  auto database_object =
+      DatabaseCatalog::GetInstance()->GetDatabaseObject(database_oid, txn);
+
+  if (!database_object || database_object->database_oid == INVALID_OID) {
+    throw CatalogException("Database " + std::to_string(database_oid) +
+                           " is not found");
+  }
+
+  // Check in pg_table using txn
+  auto table_object = database_object->GetTableObject(table_oid);
+
+  if (!table_object || table_object->table_oid == INVALID_OID) {
+    // throw table not found exception and explicitly abort txn
```

Comment:
Where is the txn explicitly being aborted ? I can only see the exception being thrown or does the exception implicitly aborts the txn?

Reply:
The traffic cop will catch the exception and set the transaction status to failure. This is preferred because this way traffic cop can send the error message to the client.